File: android/src/main/kotlin/com/inappstory/inappstory_plugin/adaptors/IASGamesAdaptor.kt
package com.inappstory.inappstory_plugin.adaptors

import IASGamesHostApi
import com.inappstory.inappstory_plugin.callbacks.GameReaderCallbackAdaptor
import com.inappstory.sdk.core.api.IASGames
import io.flutter.embedding.engine.plugins.FlutterPlugin

class IASGamesAdaptor(
    private val flutterPluginBinding: FlutterPlugin.FlutterPluginBinding,
    private val iasGames: IASGames,
) : IASGamesHostApi {
    init {
        IASGamesHostApi.setUp(flutterPluginBinding.binaryMessenger, this)

        iasGames.callback(GameReaderCallbackAdaptor(flutterPluginBinding))
    }

    override fun openGame(gameId: String) {
        iasGames.open(flutterPluginBinding.applicationContext, gameId);
    }

    override fun closeGame() = iasGames.close()

    override fun preloadGames() {
        iasGames.preloadGames();
    }


}

File: ios/inappstory_plugin/Sources/inappstory_plugin/Adaptors/IASGamesAdaptor.swift
import Flutter
import Foundation
@_spi(QAApp) import InAppStorySDK
@_spi(IAS_API) import InAppStorySDK

class IASGamesAdaptor: IASGamesHostApi {
    init(
        binaryMessenger: FlutterBinaryMessenger,
        gamesApi: InAppStorySDK.GamesAPI
    ) {
        self.binaryMessenger = binaryMessenger
        self.gamesApi = gamesApi

        IASGamesHostApiSetup.setUp(binaryMessenger: binaryMessenger, api: self)
    }

    private var binaryMessenger: FlutterBinaryMessenger

    private var gamesApi: InAppStorySDK.GamesAPI

    func openGame(gameId: String) throws {

        gamesApi.openGame(with: Game(id: gameId)) { _ in }
    }

    func closeGame() throws {
        gamesApi.closeGame()
    }

    func preloadGames() throws {
        InAppStory.shared.preloadGames()
    }
}
